### ISSUE TITLE
[Feature Request] allow opt-out of workflow isolation

### ISSUE THREAD

**Issue by neelance** [temporalio/sdk-typescript#1557]
The bug #1432 just caused a lot of pain for us. It made our workers consume a lot of memory over time (and go OOM repeatedly) and made them go really slow over time, since every (implicit) call to `new Promise` was affected. It took me several weeks of investigating, giving up and then trying again on a different day to track down this root cause. This is because the issue only happened in production and most metrics that I could get from production did not clearly point to the issue. Even diffs on heap snapshots did not clearly point to `AsyncLocalStorage` because it contained a huge amount of other heap differences and `AsyncLocalStorage` did not catch my attention. I only got lucky that a CPU profile seemed odd and I noticed that the internals of `Promise` took much too long.

Our code was using `new AsyncLocalStorage` as a constant at the toplevel of a module. Normally such a constant only gets initialised once. It is quite unintuitive that with workflow isolation one has to consider that this code gets executed multiple times and thus reason through performance implications. #1432 plans to hide these implications, but it also mentions that `maxCachedWorkflow` might still be a performance issue.

This is not the first issue that we have due to the workflow isolation feature and I'm sorry, but I doubt that it will be the last. In theory workflow isolation seems like a good idea, but in practice it adds complexity that causes very uncommon issues. I want to point out that for example the Go SDK does not try to apply workflow isolation. Instead it expects the developer to write proper workflow code that respects deterministic execution requirements. I would rather teach my team to handle a transparent nondeterminism-error due to bad workflow code than to have issues like the one above. Therefore I would like to request the option to opt-out of workflow isolation (and instead opt-in to simplicity). I still appreciate the effort you've put into workflow isolation and its performance optimisations like reusable VMs.

**Comment by mjameswh** [comment]
> ```
> TypeError: Symbol.asyncDispose is not defined.
> ```
>
> I guess https://github.com/temporalio/sdk-typescript/pull/1605 is throwing away our Symbol.asyncDispose polyfill […].

Yeah, that's very likely related, though the symptoms are exactly those I would expect. How are you loading the polfill?

I have work in progress to formalize the proper way to load polyfills; see #1743. The tricky thing is that polyfills that modify built-in globals need to be loaded early, when the workflow bundle gets loaded into the VM context (once, at worker startup time), rather than when your `workflow.ts` module actually gets executed (on every workflow start).

**Comment by mjameswh** [comment]
> This makes sense. Still, this feature could be used without all the isolation efforts, couldn't it?

Of course, and that's already supported. Just add `reuseV8Context: false` to your Worker options.

A new VM context gets created for each Workflow execution, so there's no for context bookeeping and switching. That's actually how the sandbox was initially designed, with a much simpler execution model, and therefore, less surprises and complications. The "Reuse V8 Context" optimization came in later, resulting in 2/3 reduction in memory usage and between 1/3 and 1/2 reduction in CPU usage, according to our internal stress tests and confirmed by many users. That's why we made that execution model the default, but the original executor was kept around, for those cases where restrictions of the "Reuse V8 Context" executor are too cumbersome.

**Comment by neelance** [comment]
> Once again, I fully sympathize. But exploring alternative execution models is simply not something we can prioritize at this time.

Thanks. I appreciate that you take the time for this conversation. I'll still keep a log of the issues that we see in this thread, so it can be used as a reference. It is up to you to decide which way to go.
